cmd/sonobuoy/app/pluginCache.go

```diff
@@ -0,0 +1,266 @@
+		Short: "Print the full definition of the named plugin file",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return showInstalledPlugin(getPluginCacheLocation(cmd), filenameFromArg(args[0], ".yaml"))
+		},
+	}
+
+	installCmd := &cobra.Command{
+		Use:   "install <save-as-filename> <source filename or URL>",
+		Short: "Install a plugin so that it can be run via just its filename rather than a full path or URL.",
+		Args:  cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return installPlugin(getPluginCacheLocation(cmd), filenameFromArg(args[0], ".yaml"), args[1])
+		},
+	}
+
+	uninstallCmd := &cobra.Command{
+		Use:   "uninstall <plugin filename>",
+		Short: "Uninstall a plugin. You can continue to run any plugin via specifying a file or URL.",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return deletePlugin(getPluginCacheLocation(cmd), filenameFromArg(args[0], ".yaml"))
+		},
+	}
+
+	cmd.AddCommand(listCmd, showCmd, installCmd, uninstallCmd)
+
+	return cmd
+}
+
+// getPluginCacheLocation will return the location of the plugin cache (defaults to ~/.sonobuoy)
+// If no override is set in the env, the default is assumed. If we are unable to expand to an abosolute
+// path then we return the empty string signalling the feature should be disabled.
+func getPluginCacheLocation(cmd *cobra.Command) string {
+	usePath := os.Getenv(SonobuoyDirEnvKey)
+	if len(usePath) == 0 {
+		usePath = defaultSonobuoyDir
+	}
+	expandedPath, err := homedir.Expand(usePath)
+	if err != nil {
+		logrus.Errorf("failed to expand sonobuoy directory %q: %v", usePath, err)
+		return ""
+	}
+
+	if _, err := os.Stat(expandedPath); err != nil && os.IsNotExist(err) {
+		logrus.Debugf("sonobuoy plugin location %q does not exist, creating it.", expandedPath)
+		if err := os.Mkdir(expandedPath, 0777); err != nil {
+			logrus.Errorf("failed to create directory for installed plugins %q: %v", expandedPath, err)
+		}
+	}
+
+	logrus.Debugf("Using plugin cache location %q", expandedPath)
+	return expandedPath
+}
+
+func listInstalledPlugins(installedDir string) error {
+	pluginFiles, err := loadPlugins(installedDir)
+	if err != nil {
+		return errors.Wrap(err, "failed to load installed plugins")
+	}
+
+	prefix := ""
+	first := true
+	for filename, p := range pluginFiles {
+		if !first {
+			prefix = "---\n"
+		}
+		fmt.Printf("%vfilename: %v\nplugin name: %v\nsource URL: %v\ndescription: %v\n",
+			prefix, filename, p.SonobuoyConfig.PluginName, p.SonobuoyConfig.SourceURL, p.SonobuoyConfig.Description)
+		first = false
+	}
+
+	return nil
+}
+
+// loadPlugins loads all plugins from the given directory (recursively) and
+// returns a map of their filename and contents. Returns the first error encountered
+// but continues to load and return as many plugins as possible.
+func loadPlugins(installedDir string) (map[string]*manifest.Manifest, error) {
```

Comment:
Suggestion: maybe this should return a `[]error` to collect all errs as they happen (since the walk continues). Not sure if that would be useful for the caller of this func .

Reply:
I'm always a little split on that but I tend to usually prefer the 'returns first error' since programatically, in most cases, we just need to know error/pass and the user still gets logs for each error. Makes it easier/prettier to handle rather than a []error to handle.